contracts/AINFT721.sol

```diff
@@ -0,0 +1,304 @@
+
+            
+        } else {
+            //FIXME: the originNFT contract address should not be ROLED
+            _grantRole(DEFAULT_ADMIN_ROLE, tx.origin);
+            _grantRole(PAUSER_ROLE, tx.origin);
+            _grantRole(MINTER_ROLE, tx.origin);
+        }
+        ORIGIN_NFT = IERC721(originNFT_);
+        IS_CLONED = isCloned_;
+        require((address(ORIGIN_NFT) == address(0) && !IS_CLONED) ||
+                (address(ORIGIN_NFT) != address(0) && IS_CLONED), 
+                "If AINFT721 is created first time, the originNFT_ should not be SET. \
+                \n If AINFT721 is cloned by something, the originNFT_ should be SET.");
+    }
+
+    modifier Cloned () {
+        require(IS_CLONED, "Only cloned contract can execute.");
+        _;
+    }
+    
+    function mintFromOriginInstance(
+        uint256 tokenId_
+    ) public Cloned {
+        require(!_exists(tokenId_), "The tokenId_ is already minted or cloned");
+        address originOwner = ORIGIN_NFT.ownerOf(tokenId_);
+        require(_msgSender() == originOwner, "The sender should be the holder of origin NFT.");
+        _safeMint(_msgSender(), tokenId_);
+
+    }
+    
+    function mintFromOriginInstanceOnBehalf(
+        uint256[] calldata tokenIds_,
+        address[] calldata recipients_
+    ) public Cloned onlyRole(MINTER_ROLE) {
+        for (uint i = 0; i < tokenIds_.length; i++) {
+            require(!_exists(tokenIds_[i]), "The tokenId_ is already minted or cloned");
+            require(recipients_[i] == ORIGIN_NFT.ownerOf(tokenIds_[i]), "The sender should be the holder of origin NFT.");
+            _safeMint(recipients_[i], tokenIds_[i]);
+        }
+    }
+
+    function setPaymentContract(address paymentPlugin_) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(_msgSender() == tx.origin && _msgSender() != address(0), "Only EOA can set payment contract.");
+        PAYMENT_PLUGIN = paymentPlugin_;
+    }
+
+    /**
+     * @dev See {IERC4906}.
+     */
+    function supportsInterface(
+        bytes4 interfaceId
+    )
+        public
+        view
+        override(
+            AccessControl,
+            ERC721Enumerable,
+            ERC721,
+            IERC165
+        )
+        returns (bool)
+    {
+        // IERC4906 interface added
+        return interfaceId == bytes4(0x49064906) || super.supportsInterface(interfaceId);
+    }
+
+    function pause() public onlyRole(PAUSER_ROLE) {
+        _pause();
+    }
+
+    function unpause() public onlyRole(PAUSER_ROLE) {
+        _unpause();
+    }
+
+    function safeMint(
+        address to,
+        uint256 tokenId
+    )
+        public
+        onlyRole(MINTER_ROLE)
+    {
+        //FIXME: If you use ERC721Enumerable, the original design is mint it enumerically.
+        _tokenIdCounter.increment();
+        _safeMint(to, tokenId);
+    }
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId,
```

Comment:
Urgent: `tokenId` -> `firstTokenId` : https://docs.openzeppelin.com/contracts/4.x/api/token/erc721#ERC721-_beforeTokenTransfer-address-address-uint256-uint256-

Reply:
mint checking? isCloned checking? ...

Reply:
For the cloned scenario, do we track the transferEvent of the original nft(tokenId)'s ownership?
What if it already diverged from the ownership between original's and (cloned)ainft's?

Reply:
Once the certain number of AINFT721 is minted, there is no connection between original nft and AINFT. Tracking ownership only affects in the moment of calling `mintFromOriginInstance(tokenId)`.

Thus, we don't need to track for the ownership of original NFT for the cloned scenario.
Also, the owner of original NFT and the owner of AINFT can be different after minting AINFT & transfer others.

Reply:
As we discussed above the necessity of `ERC721Enumerable`, I have a time to think which pre-checkings are needed inside `_beforeTokenTransfer()`. So in this time I'll leave it to future work.

Reply:
> Once the certain number of AINFT721 is minted, there is no connection between original nft and AINFT. Tracking ownership only affects in the moment of calling mintFromOriginInstance(tokenId).
> 
> Thus, we don't need to track for the ownership of original NFT for the cloned scenario.
Also, the owner of original NFT and the owner of AINFT can be different after minting AINFT & transfer others.

got it. I just wonder we already considered it or not.